rune/libcontainer/nsenter/loader.c

```diff
@@ -0,0 +1,123 @@
+#define _GNU_SOURCE
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <dlfcn.h>
+
+/* Defined in nsexec.c. */
+
+#define PANIC   "panic"
+#define FATAL   "fatal"
+#define ERROR   "error"
+#define WARNING "warning"
+#define INFO    "info"
+#define DEBUG   "debug"
+
+void write_log_with_info(const char *level, const char *function, int line, const char *format, ...);
+
+#define write_log(level, fmt, ...) \
+	write_log_with_info((level), __FUNCTION__, __LINE__, (fmt), ##__VA_ARGS__)
+
+struct pal_attr_t {
+	const char *args;
+	const char *log_level;
+};
+
+struct pal_stdio_fds {
+	int stdin, stdout, stderr;
+};
+
+int (*fptr_pal_get_version)(void);
+int (*fptr_pal_init)(const struct pal_attr_t *attr);
+int (*fptr_pal_exec)(const char *path, const char * const argv[],
+			const struct pal_stdio_fds *stdio, int *exit_code);
+int (*fptr_pal_kill)(int sig, int pid);
+int (*fptr_pal_destroy)(void);
+
+#define PAL_SO_PREFIX "liberpal-"
+#define PAL_SO_SUFFIX ".so"
+
+int pal_preload(void)
+{
+	char funcname[64];
+	const char *file, *basename, *suffix, *name;
+	int namelen;
+	const char *rootfs;
+	void *dl;
+
+	file = getenv("_LIBCONTAINER_PAL_PATH");
+	if (file == NULL || *file == '\0') {
+		write_log(DEBUG, "Invalid environ '_LIBCONTAINER_PAL_PATH'");
+		return -EINVAL;
+	}
+	write_log(DEBUG, "_LIBCONTAINER_PAL_PATH = %s", file);
+
+	/* fetch basename */
+	basename = strrchr(file, '/');
+	if (basename)
+		basename += 1;  /* skip '/' */
+	else
+		basename = file;
+
+	/* check prefix and suffix */
+	if (strncmp(basename, PAL_SO_PREFIX, sizeof(PAL_SO_PREFIX) - 1) != 0)
+		return -ESRCH;
+	suffix = basename + strlen(basename) - sizeof(PAL_SO_SUFFIX) + 1;
+	if (strncmp(suffix, PAL_SO_SUFFIX, sizeof(PAL_SO_SUFFIX) - 1) != 0)
+		return -ESRCH;
+
+	/* pal name */
+	name = basename + sizeof(PAL_SO_PREFIX) - 1;
+	namelen = strlen(name) - sizeof(PAL_SO_SUFFIX) + 1;
+
+	/* dlopen */
+	rootfs = getenv("_LIBCONTAINER_PAL_ROOTFS");
```

Comment:
Use signer mode to handle different paths instead of using _LIBCONTAINER_PAL_ROOTFS.